### ISSUE TITLE
`app.showRoutes()` has duplicate routes when middleware is used

### ISSUE THREAD

**Issue by Skye-31** [honojs/hono#676]
Example code
```js
app.get("/a", someMiddleWare, (c) => { 
  new Response("a");
})
app.showRoutes();
```
app.showRoutes() outputs:
```
GET    /a
GET    /a
```

**Comment by yusukebe** [comment]
Hi @Skye-31 !

Thank you for creating the issue.

It is not a problem that `app.showRoutes()` shows duplicate routes. `app.showRoutes()` just displays the registered routes in order. When displaying, it is possible to remove duplicates, but in this way, we can know how many Handlers or Middleware are registered. I think this is a better way.

**Comment by Skye-31** [comment]
It could be neater to display it as the following, or something similar? 🙂 
```
GET    /a (2x)
```

**Comment by Skye-31** [comment]
I just thought I'd raise it since it seemed like a bug - I'm not too bothered if it overcomplicates things. Just found it confusing for my app to display a few hundred routes when I'm actually only running about 70 😅

**Comment by dario-piotrowicz** [comment]
Would it be possible to add a parameter for the `showRoutes` method not to show duplicate routes if that's what the developer wants?

**Comment by yusukebe** [comment]
> Would it be possible to add a parameter for the showRoutes method not to show duplicate routes if that's what the developer wants?

Yes. But I don't want to give the options to developers for the `showRoutes` to make this simple. And it may make the code verbose mentioned above. In this case, the problem is the developer may feel, "Is this bug?". 

For example, there could be a more clear way to show the routes, such as changing the color, numbering, and so on. However, we don't know the best way.

**Comment by dagnelies** [comment]
Hi, here again a little feedback concerning the representation.
Although I understand the output of showRoutes() I was a little bit confused by the routes not representing the endpoints.

I personally think the ideal representation would be something like this: 

```
GET /a   [logger, basicAuth]
GET /b   [logger, cors, bearerAuth]
```

Or alternatively at least:

```
> ALL *  [logger]
> ALL /a  [basicAuth]
GET /a
> GET /b  [cors]
> ALL /b   [bearerAuth]
GET /b
```

Just to distinguish better which routes are middleware and which routes are actual endpoints. Dunno if it's possible to have named middleware show up there, it would be the bonus. :)

**Comment by yusukebe** [comment]
> I personally think the ideal representation would be something like this:
> Just to distinguish better which routes are middleware and which routes are actual endpoints. 

Yes, of course, I also want to do that, but I could not implement it like that, I think there is no way:(

**Comment by transitive-bullshit** [comment]
Found this issue thinking it was a bug. The current `showRoutes` is really confusing, since middleware shouldn't be considered routes. I'd much rather have this function just show the list of routes without any numbers or middleware considerations.

**Comment by yusukebe** [comment]
@transitive-bullshit

In Hono, middleware and handlers are treated the same. "Routes" in Hono include middleware. Whether something is a "handler" or "middleware" is determined at the time of handling the first access. Therefore, when `app.showRoutes()` is executed, it's not possible to distinguish whether it's a handler or middleware. Additionally, using `app.use()` or `app.HTTP_METHOD()` doesn't necessarily indicate whether it's middleware or not.

Although, seeing all handlers and middleware could indeed be very helpful.

But, if you felt this as a bug, we may need to consider addressing it.

**Comment by usualoma** [comment]
Although not perfectly distinguishable, middleware and handlers usually differ in the number of arguments.

```ts
const middleware = async (c, next) => { await next() }
const handler = (c) => c.text('hi')
```

The number of arguments can be obtained with `.length`.

<img width="214" alt="image" src="https://github.com/honojs/hono/assets/30598/f4a8a252-b76e-4f6a-ad24-2438c5eb971e">

There is some debate about whether this can be used as run-time information. (It may be that `(. .args)` might be used in some cases. Seems unlikely ;-)

However, even if it cannot be used for runtime information, it may be a hint for `showRoutes()` to work.

### Should showRoutes() be a method of an instance of Hono?

Since this function is used during development, it should be provided as a development tool, not loaded in the production environment.
How about doing it this way?
https://github.com/honojs/hono/pull/1697

**Comment by transitive-bullshit** [comment]
I'd be in favor of moving it to a utility method.

**Comment by yusukebe** [comment]
Yeah, making it as a "dev helper" is a good idea.

@usualoma 

Let's continue our discussion on issue #1697.

**Comment by yusukebe** [comment]
This issue is resolved by #1716 .
